test/e2e/runner/infra.go

```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"context"
+
+	e2e "github.com/tendermint/tendermint/test/e2e/pkg"
+)
+
+// TestnetInfra provides an API for interacting with the infrastructure for an
+// entire testnet.
+type TestnetInfra interface {
+	// GenerateConfig generates any necessary configuration for the
+	// infrastructure provider during testnet setup.
+	GenerateConfig() error
+
+	// Stop attempts to stop the entire testnet.
+	Stop(ctx context.Context) error
+
+	// Pause attempts to pause the entire testnet.
+	Pause(ctx context.Context) error
+
+	// Unpause attempts to resume a paused testnet.
+	Unpause(ctx context.Context) error
+
+	// ShowLogs will print all logs for the whole testnet to stdout.
+	ShowLogs(ctx context.Context) error
+
+	// TailLogs tails the logs for all nodes in the testnet.
+	TailLogs(ctx context.Context) error
+
+	// Cleanup stops and destroys all running infrastructure and deletes any
+	// generated files.
+	Cleanup(ctx context.Context) error
+}
+
+// NodeInfra provides an API for interacting with specific nodes'
+// infrastructure.
+type NodeInfra interface {
+	// ProvisionNode attempts to provision infrastructure for the given node
+	// and starts it.
+	ProvisionNode(ctx context.Context, node *e2e.Node) error
```

Comment:
I think the NodeInfra interface methods don't need the word "Node" in the method name.

Reply:
> I'm still not convinced that the testnet implementation needs access to the internals of the node implementation, except as an optimization, is this correct?

Unfortunately not, the runner CLI provides commands by which users can manipulate individual nodes. I won't be able to reduce the surface area of the new interface without substantially reducing the functionality provided by the testnet runner at present, and I'm not sure what the unintended consequences of that will be. Perhaps we can set that as a longer-term goal once we've implemented the Digital Ocean-based infra API?

Reply:
>  the runner CLI provides commands by which users can manipulate individual nodes

This might actually be ok to remove. We should check with @cmwaters, but I don't think that any of these CLI operations are used for anything, except in theory making it easier to avoid reaching into the provider directly (which you can always do, and maybe ought to in these cases.)

Reply:
> the runner CLI provides commands by which users can manipulate individual node

Which commands are these - I thought all commands were for an entire testnet